### PR TITLE
lregex: use pcre2 if p flag is given

### PR DESCRIPTION
**Pull request by masatake**
Signed-off-by: Masatake YAMATO <yamato@redhat.com>

- [ ] as a testbed, add a NEW parser using pcre2 aggressively. ~I'm targeting jinja2.~
- [ ] add test cases for mtable and mline. ~The jinja2 may be a mtable parser.~
- [x] revise flags passed to pcre2_compile. read #1861.
- [x] introduce a thin layer that hides regex engines from the main part.
- [x] ~support the engine in optlib2c.~ This change is nothing to do with optlib2c.
- [x] don't allocate 'match_data' data structure every time when matching.  Pre-allocate one and reuse it.

Optional tasks:
- [ ] add code to link onigruma. It seems that pcre2 doesn't include akr's works that I would like to use.
- [ ] provide the way to access named group from optscript. How? `\/name` ?
- [ ] add --regengine-<LANG>=pcre2 or something an option to force using pcre2 as default. This is useful to reuse existing test cases to test pcre2 backend. I cannot find a good name for the option.

### CODE REVIEW

File: win32/appveyor.bat [hunk 1]
```diff
@@ -139,7 +139,7 @@ if "%normalbuild%"=="no" (
   bash -lc "for i in {1..3}; do pacman --noconfirm --noprogressbar -S --needed mingw-w64-%MSYS2_ARCH%-python3-sphinx && break || sleep 15; done"
 )
 :: Install necessary packages
-bash -lc "for i in {1..3}; do pacman --noconfirm --noprogressbar -S --needed mingw-w64-%MSYS2_ARCH%-{jansson,libxml2,libyaml} && break || sleep 15; done"
+bash -lc "for i in {1..3}; do pacman --noconfirm --noprogressbar -S --needed mingw-w64-%MSYS2_ARCH%-{jansson,libxml2,libyaml,pcre} && break || sleep 15; done"
```
Comment (k-takata) [inline win32/appveyor.bat:142]:
```suggestion
bash -lc "for i in {1..3}; do pacman --noconfirm --noprogressbar -S --needed mingw-w64-%MSYS2_ARCH%-{jansson,libxml2,libyaml,pcre2} && break || sleep 15; done"
```
maybe...

Comment (masatake) [reply win32/appveyor.bat:142]: Thank you! I will update my change.

**Comment by masatake** [comment]
I modified test cases to work them even with pcre2. I think the changes are enough minor and acceptable.

The default engine is "e'. "e" specifies the regex engine we are using now.
```
diff --git a/main/lregex.c b/main/lregex.c
index cdadca1c0..16bb1529c 100644
--- a/main/lregex.c
+++ b/main/lregex.c
@@ -63,7 +63,7 @@ static bool regexAvailable = false;
    Tmain cases. */
 #define MTABLE_MOTIONLESS_MAX (MTABLE_STACK_MAX_DEPTH + 1)
 
-#define DEFAULT_REGEX_BACKEND "e"
+#define DEFAULT_REGEX_BACKEND "p"
 
 /*
 *   DATA DECLARATIONS
```

You can switch the default engine to pcre2 with the above change.
I had a plan to provide a way to switch the default engine from cmdline.

PuppetManifest is a parser using multi-table regex aggressively.
I evaluated "e" and "p" with the parser. https://github.com/universal-ctags/codebase provides inputs:

With "e" engine:
```
[yamato@control]~/var/codebase% ./codebase ctags PuppetManifest                                       
version: 883d05c6
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +packcc +optscript +pcre2
log: results/883d05c6,PuppetManifest......,..........,time......,default...,2021-05-27-23:25:38.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=PuppetManifest -o - -R code/puppet-examples code/puppet code/theforeman-puppet code/wikimedia-puppet
u-ctags: Warning: cannot open input file "code/wikimedia-puppet/modules/admin/files/home/akosiaris/.my.cnf" : Permission denied
2681 files, 123626 lines (4136 kB) scanned in 1.1 seconds (3727 kB/s)
23950 tags added to tag file

real	0m1.139s
user	0m1.075s
sys	0m0.060s
+ set +x
```

With "p" engine:
```
[yamato@control]~/var/codebase% ./codebase ctags PuppetManifest
version: 883d05c6
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +packcc +optscript +pcre2
log: results/883d05c6,PuppetManifest......,..........,time......,default...,2021-05-27-23:26:19.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=PuppetManifest -o - -R code/puppet-examples code/puppet code/theforeman-puppet code/wikimedia-puppet
u-ctags: Warning: cannot open input file "code/wikimedia-puppet/modules/admin/files/home/akosiaris/.my.cnf" : Permission denied
2681 files, 123626 lines (4136 kB) scanned in 9.9 seconds (417 kB/s)
23950 tags added to tag file

real	0m9.955s
user	0m9.881s
sys	0m0.058s
+ set +x
```

About 9 times slower.  Disappointed. I think I made some mistakes in the code using pcre2.

**Comment by masatake** [comment]
```
Samples: 40K of event 'cycles:uppp', Event count (approx.): 36867296116
  Children      Self  Command  Shared Object        Symbol
-   89.60%    89.59%  u-ctags  libpcre2-8.so.0.8.0  [.] _pcre2_strlen_8
     _pcre2_strlen_8
+    4.17%     4.17%  u-ctags  libpcre2-8.so.0.8.0  [.] pcre2_match_8
+    3.62%     0.00%  u-ctags  [unknown]            [.] 0x00000000801f0fff
+    3.62%     0.00%  u-ctags  libpcre2-8.so.0.8.0  [.] default_free
+    0.76%     0.76%  u-ctags  libc-2.27.so         [.] __memset_avx2_unaligned_erms
+    0.69%     0.00%  u-ctags  [unknown]            [.] 0000000000000000
+    0.65%     0.00%  u-ctags  [unknown]            [.] 0x0000000001ded3a0
+    0.62%     0.00%  u-ctags  [unknown]            [.] 0x0000000001e31810
```

perf says strlen alike function called frequently. This explains "9 times slower" well.
Who calls strlen so many times?

**Comment by masatake** [comment]
I found the way to remove strlen calls:
Now "pcre2" is:
```
[yamato@control]~/var/codebase% ./codebase ctags PuppetManifest       
version: 883d05c6
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +packcc +optscript +pcre2
log: results/883d05c6,PuppetManifest......,..........,time......,default...,2021-05-28-00:27:23.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --totals=yes --languages=PuppetManifest -o - -R code/puppet-examples code/puppet code/theforeman-puppet code/wikimedia-puppet
u-ctags: Warning: cannot open input file "code/wikimedia-puppet/modules/admin/files/home/akosiaris/.my.cnf" : Permission denied
2681 files, 123626 lines (4136 kB) scanned in 0.8 seconds (5091 kB/s)
23950 tags added to tag file

real	0m0.840s
user	0m0.789s
sys	0m0.048s
+ set +x
```

**Comment by masatake** [comment]
92166be shows the difference of characters of regex engines.

**Comment by masatake** [comment]
I would like to withdraw this pull request.
The parser I would like to write can be written with posix regex + optscript.
A parser written in optscript may be 10 ~ 30 times slower than pcre2 based parser.
However, my parsers are for rather small input. So the performance may not be issue.

If you are interested in taking this pull request, consider adding a new parser using pcre2.

jinja2 parser was my target during developing the code of this pull request, introducing pcre2.
For extracting parameter lists (signatures) from macros, I used `?R` operator of pcre2.
However, I found this can be done with optscript. With pcre2, we can extract the signature in very short code.
However, for normalizing the signature, eventually we need a code written in optscript.
Removing duplicated spaces and/or replacing a newline with a space are an example of the normalization.
 
The changes from b3b3b41 to aa7b7a0 have worth for merging even if we don't introduce pcre2.
So I will make a new pull for them.

```
#
#  jinja2.ctags --- a parser for jinja2 templates
#
#  Copyright (c) 2021 Masatake YAMATO
#  Copyright (c) 2021 Red Hat, Inc.
#
#  This source code is released for free distribution under the terms of the
#  GNU General Public License version 2 or (at your option) any later version.
#
# References:
#
# - https://jinja.palletsprojects.com/en/3.0.x/templates/
#

--langdef=Jinja2

#
# map
#
--map-Jinja2=+.j2
--map-Jinja2=+.jinja2

#
# kinds
#
--kinddef-Jinja2=m,macro,macros

#
# table definitions
#
--_tabledef-Jinja2=preproc
--_tabledef-Jinja2=main
--_tabledef-Jinja2=directive
--_tabledef-Jinja2=comment
--_tabledef-Jinja2=skipToEnd


#
# table implementation
#

# The table for running a guest
--_mtable-regex-Jinja2=preproc/.*//{_guest=.basename,0start,0end}{_advanceTo=0start}{tjump=main}

# The main table
--_mtable-regex-Jinja2=main/[^\{]+//
--_mtable-regex-Jinja2=main/\{#//{tenter=comment}
--_mtable-regex-Jinja2=main/\{%[- \t\n]+//{tenter=directive}
--_mtable-regex-Jinja2=main/.//

# Handling directives
--_mtable-regex-Jinja2=directive/macro ([a-zA-Z_0-9.]+)[ \t]*(\((?:[^()]+|(?R))*\))/\1/m/{pcre2}{tjump=skipToEnd}{{
    . dup \2 signature:
    % Push the tag entry for filling end: field with endmacro.
}}

--_mtable-regex-Jinja2=directive/(endmacro)//{tjump=skipToEnd}{{
    1 _matchloc end:
}}

# import, set, block, extends
--_mtable-regex-Jinja2=directive/.//

# Handling the end of directives
--_mtable-regex-Jinja2=skipToEnd/[^%]+//
--_mtable-regex-Jinja2=skipToEnd/%\}//{tleave}

# Skip comments
--_mtable-regex-Jinja2=comment/[^#]+//
--_mtable-regex-Jinja2=comment/#\}//{tleave}
--_mtable-regex-Jinja2=comment/.//
```

**Comment by hirooih** [comment]
@k-takata san,

> See Akira Tanaka's original paper (in Japanese) for detail:
> https://staff.aist.go.jp/tanaka-akira/pub/prosym49-akr-paper.pdf

Thank you.  I read this paper.
It proposes the absence operator `(?~)` (`(?!)` in the paper).

`<-(?~->)->`

The paper shows two methods implemented in PCRE as known equivalent works.

1. Preventing backtracking:  `<-(?>.*?->)`

[Using independent subexpressions to prevent backtracking](https://perldoc.perl.org/perlretut#Using-independent-subexpressions-to-prevent-backtracking) in Perl regular expressions tutorial

2. negative lookahead assertion: `/<-((?!->).)*->`

[Looking ahead and looking behind](https://perldoc.perl.org/perlretut#Looking-ahead-and-looking-behind) in Perl regular expressions tutorial

As a user, I don't care that the absence operator "has proper semantics on the theory", but it is much more easy to use.

Thank you for giving me the chance to study regular expressions more.

**Pull request closed without merging**